### PR TITLE
Added WhereLine and "=" edit command

### LINKED ISSUE #97: Implement "=" editor command
Currently we only support "=#", which will just "Print just the character address of the range." "=" should "Print the line address and character address of the range."

### PR DESCRIPTION
**Pull request by leGoof**
WhereLine will return the line address of a given address.
The "=" edit command will return the line(s) of an address.
This command differs slightly from sam which returns both the line(s)
and the character address.

Fixes #97.

### CODE REVIEW

File: edit/edit.go [hunk 1]
```diff
@@ -290,6 +290,45 @@ func where(ed *Editor, a Address) (addr, error) {
 	return at, nil
 }
 
+// WhereLine returns the line number corresponding to the start and
+// end of the address.
+// Dot is set to the address.
+func (ed *Editor) WhereLine(a Address) (lineFrom, lineTo int64, err error) {
```
Comment (eaburns) [inline edit/edit.go]: Just use int for the line numbers. They will never be big enough to need int64.

File: edit/edit.go [hunk 2]
```diff
@@ -290,6 +290,45 @@ func where(ed *Editor, a Address) (addr, error) {
...
+		if err != nil {
+			return 0, 0, err
+		} else if r == '\n' {
+			from++
+		}
+	}
+	to = from
+	i++
+	for ; i < at.to; i++ {
+		r, err := ed.buf.rune(i - 1)
```
Comment (eaburns) [inline edit/edit.go]: If you didn't pre-increment i on line 319, could you avoid the -1 here?

File: edit/edit.go (hunk 1 above)
Comment (eaburns) [inline edit/edit.go]: I dunno. Maybe it's fine. You pick. ints are nicer to use, but int64 would make it consistent with Where.

File: edit/edit.go [hunk 3]
```diff
@@ -619,14 +659,25 @@ func edit(ed *Editor, cmd []rune) ([]rune, addr, error) {
...
-		if cmd[1] != '#' {
-			return nil, addr{}, errors.New("unknown command: " + string(c))
-		}
+		var ret string
 		at, err := where(ed, a)
 		if err != nil {
 			return nil, addr{}, err
 		}
-		return []rune(fmt.Sprintf("#%d,#%d", at.from, at.to)), at, nil
+		if len(cmd) == 1 || cmd[1] != '#' {
```
Comment (eaburns) [inline edit/edit.go]: Can we move the logic for computing the string version (this if) into a separate function? I want to keep the cases of this switch pretty small, because it's already quite long.

File: edit/edit_test.go [hunk 4]
```diff
@@ -93,6 +93,40 @@ func (test *whereTest) run(t *testing.T) {
...
 }
 
+func TestWhereLine(t *testing.T) {
+	tests := []whereLineTest{
+		{init: "Hello\n 世界!", addr: All, from: 1, to: 2},
+		{init: "Hello\n 世界!", addr: End, from: 2, to: 2},
+		{init: "Hello\n 世界!", addr: Line(1), from: 1, to: 1},
+		{init: "Hello\n 世界!", addr: Line(2), from: 2, to: 2},
+		{init: "Hello\n 世界!", addr: Regexp("/Hello"), from: 1, to: 1},
+		{init: "Hello\n 世界!", addr: Regexp("/世界"), from: 2, to: 2},
```
Comment (eaburns) [inline edit/edit_test.go]: Can we add some tests:
* init: ""
* The address spans multiple lines.
* A more than two lines.

File: edit/edit_test.go [hunk 5]
```diff
@@ -629,6 +663,21 @@ func TestEditorEditWhere(t *testing.T) {
...
 }
 
+func TestEditorEditWhereLine(t *testing.T) {
+	tests := []editTest{
+		{
+			{edit: "a/Hello\n World!", want: "Hello\n World!"},
+			{edit: "=", print: "1,2", want: "Hello\n World!"},
+			{edit: "/e.*\n.*/=", print: "1,2", want: "Hello\n World!"},
+			{edit: "0=", print: "1", want: "Hello\n World!"},
+			{edit: "$=", print: "2", want: "Hello\n World!"},
```
Comment (eaburns) [inline edit/edit_test.go]: Can we add the same tests as I suggested above?

File: edit/edit.go (hunk 2 above)
Comment (leGoof) [inline edit/edit.go]: Hrm... I think so, but if we did that we'd also need to do "i < at.to -1".

Comment (eaburns) [inline edit/edit.go]: Why don't you want to check the rune at `at.to-1`? That rune is within the address too.
We need to add some more tests:
* where the selection begins with \n
* where the selection ends with \n
* where the selection {begins, ends} just {after,before} \n

I want to make sure that we correctly handle these edge cases.

Comment (eaburns) [inline edit/edit.go]: I see, it's because if `at.to-1` is `\n` then you don't want to count the extra newline. If `at.to-1` isn't `\n` then the address ends on the current line and `to` is already correct. But that's pretty subtle, in my opinion.

How about making it more explicit (even if it's a bit redundant):
```go
to = from
for ; i < at.to; i++ {
  r, err := ed.buf.rune(i)
  if err != nil {
    return 0, 0, err
  } else if r == '\n' && i != at.to-1 {
    to++
  }
}
```

Would that work?

File: edit/edit.go (hunk 1 above)
Comment (leGoof) [inline edit/edit.go]: I'll go with int's.

File: edit/edit.go (hunk 2 above)
Comment (leGoof) [inline edit/edit.go]: Works for me.

File: edit/edit.go (hunk 3 above)
Comment (leGoof) [inline edit/edit.go]: Ehhh, that would really only make the edit function 5 lines shorter (the "if lineFrom == lineTo ...") I don't think it's worth it to move it to its own function, especially since we need to add the same check to the "=#" command.

File: edit/edit_test.go (hunk 4 above)
Comment (leGoof) [inline edit/edit_test.go]: Yup.

File: edit/edit_test.go (hunk 5 above)
Comment (leGoof) [inline edit/edit_test.go]: Yup.

File: edit/edit.go (hunk 1 above)
Comment (mccoyst) [inline edit/edit.go]: Addresses use int64, may as well be consistent, so that my file filled with ^int(0) newlines is editable in T.

Comment (leGoof) [inline edit/edit.go]: Big shrugs. I'll stick with int64 barring another more compelling reason to stick with just ints.

**Comment by leGoof** [comment]
I changed where to whereAddr, but upon further reflection maybe it should be whereRune? I'll hold off on any more changes here until I get more feedback.

File: edit/edit.go (hunk 3 above)
Comment (eaburns) [inline edit/edit.go]: OK.

File: edit/edit.go (hunk 1 above)
Comment (eaburns) [inline edit/edit.go]: I like int, because the `Line` function that creates a line address uses int.

File: edit/edit.go [hunk 6]
```diff
@@ -272,17 +272,21 @@ func print(ed *Editor, a Address) ([]rune, addr, error) {
...
+	at, err := whereAddr(ed, a)
+	if err != nil {
+		return 0, 0, 0, 0, err
+	}
+	lfrom, lto, err = whereLine(ed, a)
+	return at.from, at.to, lfrom, lto, err
 }
 
-func where(ed *Editor, a Address) (addr, error) {
+func whereAddr(ed *Editor, a Address) (addr, error) {
```
Comment (eaburns) [inline edit/edit.go]: Doesn't matter much to me. It's not exported. Though I do slightly prefer just `where` so that it matches `Where`. I may change some of these names when I do #103.

**Comment by eaburns** [comment]
Looks good to me. When we figure out who's job it is to merge, we should rebase and merge it.
